lib/systems/inspect.nix

```diff
@@ -1,14 +1,64 @@
 { lib }:
-with import ./parse.nix { inherit lib; };
-with lib.attrsets;
-with lib.lists;
 
-let abis_ = abis; in
-let abis = lib.mapAttrs (_: abi: builtins.removeAttrs abi [ "assertions" ]) abis_; in
+let
+  # TODO: expose this from lib.attrsets
+  inherit (builtins) intersectAttrs;
+
+  inherit (lib.attrsets)
+    attrValues
+    mapAttrs
+    matchAttrs
+    recursiveUpdateUntil
+    removeAttrs
+    ;
+
+  inherit (lib.lists) any concatMap filter isList toList;
+
+  inherit (lib.strings) hasPrefix toJSON;
+
+  inherit (lib.systems.parse)
+    kernels
+    kernelFamilies
+    significantBytes
+    cpuTypes
+    execFormats
+    ;
+
+  abis = mapAttrs (_: abi: removeAttrs abi [ "assertions" ]) lib.systems.parse.abis;
+
+  inherit (abis)
+    android
+    androideabi
+    cygnus
+    gnu
+    gnuabi64
+    gnuabielfv1
+    gnuabielfv2
+    gnuabin32
+    gnueabi
+    gnueabihf
+    musl
+    muslabi64
+    muslabin32
+    musleabi
+    musleabihf
+    uclibc
+    uclibceabi
+    uclibceabihf
+    ;
+
+  # These patterns depend on each other, so they're at the top level.
+  isBSD          = { kernel = { families = { inherit (kernelFamilies) bsd; }; }; };
+  isDarwin       = { kernel = { families = { inherit (kernelFamilies) darwin; }; }; };
+  isUnix         = [ isBSD isDarwin isLinux isSunOS isCygwin isRedox ];
+
+  isLinux        = { kernel = kernels.linux; };
+  isSunOS        = { kernel = kernels.solaris; };
+  isCygwin       = { kernel = kernels.windows; abi = cygnus; };
+  isRedox        = { kernel = kernels.redox; };
```

Comment:
Pulling these things apart makes it harder to have things in overview. I don't think `rec` per se is bad if used judiciously, and can in fact greatly improve clarity. That ["best practice" recommendation](https://nix.dev/guides/best-practices#recursive-attribute-set-rec) is an oversimplification and needs more work.

Reply:
I hear that. I find the exact semantics of `rec` really hard to articulate, but you're right that it's a separate-ish concern from `with`. Perhaps this is me suffering from the overeager schoolboy problem of trying to get the "extra" in.

Reply:
I've removed the commit that did this from this PR.